drivers/clock_control/clock_control_renesas_cpg_mssr.c

```diff
@@ -58,3 +58,319 @@ int rcar_cpg_mstp_clock_endisable(uint32_t base_address, uint32_t module, bool e
 
 	return 0;
 }
+
+struct cpg_clk_info_table *
+rcar_cpg_find_clk_info_by_module_id(const struct device *dev, uint32_t domain, uint32_t id)
+{
+	struct rcar_cpg_mssr_data *data = dev->data;
+	uint32_t left = 0, right = data->clk_info_table_size[domain] - 1;
+
+	while (left <= right) {
+		uint32_t middle = left + ((right - left) >> 1);
+		struct cpg_clk_info_table *item = &data->clk_info_table[domain][middle];
+		uint32_t module = item->module;
+
+		if (module == id) {
+			return item;
+		}
+
+		if (module < id) {
+			left = middle + 1;
+		} else {
+			right = middle - 1;
+		}
+	}
+
+	LOG_ERR("%s: can't find clk info (domain %u module %u)", dev->name, domain, id);
+	return NULL;
+}
+
+static uint32_t rcar_cpg_get_divider(const struct device *dev, struct cpg_clk_info_table *clk_info)
+{
+	mem_addr_t reg_addr;
+	mm_reg_t reg_val;
+	uint32_t divider = RCAR_CPG_NONE;
+	struct rcar_cpg_mssr_data *data = dev->data;
+
+	if (clk_info->domain == CPG_MOD) {
+		return 1;
+	}
+
+	reg_addr = clk_info->offset;
+	if (reg_addr == RCAR_CPG_NONE) {
+		/* if we don't have valid offset, in is equal to out */
+		return 1;
+	}
+
+	reg_addr += data->base_addr;
+	reg_val = sys_read32(reg_addr);
+
+	if (data->get_div_helper) {
+		divider = data->get_div_helper(reg_val, clk_info->module);
+	}
+
+	if (!divider) {
+		return RCAR_CPG_NONE;
+	}
+
+	return divider;
+}
+
+static int rcar_cpg_update_out_freq(const struct device *dev, struct cpg_clk_info_table *clk_info)
+{
+	uint32_t divider = rcar_cpg_get_divider(dev, clk_info);
+
+	if (divider == RCAR_CPG_NONE) {
+		return -EINVAL;
+	}
+
+	clk_info->out_freq = clk_info->in_freq / divider;
+	return 0;
+}
+
+static int rcar_cpg_get_in_update_out_freq(const struct device *dev,
```

Comment:
Why are you returning int? AFAIR, clock rates in zephyr are defined as `uint32_t`. 

(`uint32_t` is not the best case either, but it is completely different topic)